File: oci_iam/OCI_compartments_list_formatted.py
#!/usr/bin/env python3

# --------------------------------------------------------------------------------------------------------------
# This script lists all the compartments and sub-compartments names and IDs in an OCI tenant using OCI Python SDK
# The output is formatted with colors and indents to easily identify parents of sub-compartments
# Note: OCI tenant given by an OCI CLI PROFILE
# It is much faster than the corresponding Bash script on tenant with many compartments
#
# Author        : Christophe Pauliat
# Platforms     : MacOS / Linux
# prerequisites : - Python 3 with OCI Python SDK installed
#                 - OCI config file configured with profiles
# Versions
#    2019-10-18: Initial Version
#    2020-04-24: minor code enhancements
#    2022-01-03: use argparse to parse arguments
#    2022-01-04: add --no_color option
# --------------------------------------------------------------------------------------------------------------

# -------- import
import oci
import sys
import argparse

# -------- colors for output
COLOR_YELLOW = "\033[93m"
COLOR_RED    = "\033[91m"
COLOR_GREEN  = "\033[32m"
COLOR_NORMAL = "\033[39m"
COLOR_CYAN   = "\033[96m"
COLOR_BLUE   = "\033[94m"
COLOR_GREY   = "\033[90m"

# -------- variables
configfile = "~/.oci/config"    # Define config file to be used.
flag=[0,0,0,0,0,0,0,0,0,0]

# -------- functions

# ---- usage syntax
def usage():
    print ("Usage: {} [-nc] [-d] -p OCI_PROFILE".format(sys.argv[0]))
    print ("")
    print ("    If -d is provided, deleted compartments are also listed.")
    print ("    If not, only active compartments are listed.")
    print 
    print ("note: OCI_PROFILE must exist in {} file (see example below)".format(configfile))
    print ("")
    print ("[EMEAOSCf]")
    print ("tenancy     = ocid1.tenancy.oc1..aaaaaaaaw7e6nkszrry6d5hxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    print ("user        = ocid1.user.oc1..aaaaaaaayblfepjieoxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    print ("fingerprint = 19:1d:7b:3a:17:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx")
    print ("key_file    = /Users/cpauliat/.oci/api_key.pem")
    print ("region      = eu-frankfurt-1")
    exit (1)

# ---- Disable colored output
def disable_colored_output():
    global COLOR_YELLOW 
    global COLOR_RED   
    global COLOR_GREEN 
    global COLOR_NORMAL 
    global COLOR_CYAN   
    global COLOR_BLUE   
    global COLOR_GREY    

    COLOR_YELLOW = ""
    COLOR_RED    = ""
    COLOR_GREEN  = ""
    COLOR_NORMAL = ""
    COLOR_CYAN   = ""
    COLOR_BLUE   = ""
    COLOR_GREY   = ""

def get_cpt_name_and_state_from_id(cpt_id):
    for c in compartments:
        if (c.id == cpt_id):
            return c.name, c.lifecycle_state
    return

def list_compartments(parent_id, level):
    # level = 0 for root, 1 for 1st level compartments, ...
    Debug=False

    if (Debug):
        print ("NEW ITER: DEBUG: level=%d parent_id=%s " % (level, parent_id), end='')
        print ("flag=",flag)
    
    i=1
    while i < level:
        if flag[i] == 0:
            print (COLOR_CYAN+"│      "+COLOR_NORMAL,end='')
        else:
            print ("       ",end='')
        i += 1    

    if level > 0:
        cptname, state = get_cpt_name_and_state_from_id (parent_id)   
   
        if flag[level] == 0:
            print (COLOR_CYAN+"├───── "+COLOR_NORMAL,end='')
        else:
            print (COLOR_CYAN+"└───── "+COLOR_NORMAL,end='')
    else:
        cptname='root'
        state="ACTIVE"
    
    if state == "ACTIVE":
        print (COLOR_GREEN+cptname+COLOR_NORMAL+" "+parent_id+COLOR_YELLOW+" ACTIVE"+COLOR_NORMAL)
    else:
        print (COLOR_BLUE+cptname+COLOR_GREY+" "+parent_id+COLOR_RED+" DELETED"+COLOR_NORMAL)

    # get the list of ids of the direct sub-compartments
    sub_compartments_ids_list=[]
    for c in compartments:
        if c.compartment_id == parent_id:
            if list_deleted or c.lifecycle_state != "DELETED":
                sub_compartments_ids_list.append(c.id)
    
    # then for each of those cpt ids, display the sub-compartments details
    if (Debug):
        print ("DEBUG: len=%d" % len(sub_compartments_ids_list))
    i=1
    for cid in sub_compartments_ids_list:     
        # if processing the last sub dir
        if (Debug):
            print ("DEBUG: test child %s" % cid)  
        if i == len(sub_compartments_ids_list):
            flag[level+1]=1
        else:
            flag[level+1]=0
        if (Debug):
            print ("DEBUG: flag", flag)
        list_compartments(cid, level+1)
        i += 1

# -------- main

# -- parsing arguments
parser = argparse.ArgumentParser(description = "List compartments in an OCI tenant")
parser.add_argument("-p", "--profile", help="OCI profile", required=True)
parser.add_argument("-d", "--list_deleted", help="List also deleted compartments", action="store_true")
parser.add_argument("-nc", "--no_color", help="Disable colored output", action="store_true")
args = parser.parse_args()
    
profile       = args.profile
list_deleted  = args.list_deleted
if args.no_color:
  disable_colored_output()

# -- get OCI Config
try:
    config = oci.config.from_file(configfile,profile)
except:
    print ("ERROR: profile '{}' not found in config file {} !".format(profile,configfile))
    exit (2)

IdentityClient = oci.identity.IdentityClient(config)
user = IdentityClient.get_user(config["user"]).data
RootCompartmentID = user.compartment_id

# -- get list of compartments with all sub-compartments
response = oci.pagination.list_call_get_all_results(IdentityClient.list_compartments,RootCompartmentID,compartment_id_in_subtree=True)
compartments = response.data

list_compartments(RootCompartmentID,0)

exit (0)

File: OCI_compartments_list_formatted.go
// --------------------------------------------------------------------------------------------------------------
// This script lists the compartment names and IDs in a OCI tenant using OCI Go SDK
// The output is formatted with colors and indents to easily identify parents of sub-compartments
// Note: OCI tenant given by an OCI CLI PROFILE
// Author        : Christophe Pauliat
// Platforms     : MacOS / Linux
// prerequisites : - Go with OCI SDK installed
//                 - OCI config file configured with profiles
// Versions
//    2020-06-25: Initial Version
// --------------------------------------------------------------------------------------------------------------


package main

// -- import
import (
	"context"
	"fmt"
	"os"

	"github.com/oracle/oci-go-sdk/common"
	"github.com/oracle/oci-go-sdk/identity"
	"github.com/oracle/oci-go-sdk/example/helpers"
)

// -- constants
const config_file string = "~/.oci/config"    // Define config file to be used.
const COLOR_YELLOW = "\033[93m"
const COLOR_RED    = "\033[91m"
const COLOR_GREEN  = "\033[32m"
const COLOR_NORMAL = "\033[39m"
const COLOR_CYAN   = "\033[96m"
const COLOR_BLUE   = "\033[94m"
const COLOR_GREY   = "\033[90m"

// -- global variables
var flag [10]int

// -- functions
func usage() {
    fmt.Printf ("Usage: %s OCI_PROFILE\n",os.Args[0])
    fmt.Println("")
	fmt.Printf ("note: OCI_PROFILE must exist in %s file (see example below)\n",config_file)
	fmt.Println("")
    fmt.Println("[EMEAOSCf]")
    fmt.Println("tenancy     = ocid1.tenancy.oc1..aaaaaaaaw7e6nkszrry6d5hxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    fmt.Println("user        = ocid1.user.oc1..aaaaaaaayblfepjieoxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    fmt.Println("fingerprint = 19:1d:7b:3a:17:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx")
    fmt.Println("key_file    = /Users/cpauliat/.oci/api_key.pem")
    fmt.Println("region      = eu-frankfurt-1")
	os.Exit (1)	
}

func get_cpt_name_and_state_from_id(cpt_id string, cpts []identity.Compartment) (string, string) {
	for _, c := range cpts {
        if (*c.Id == cpt_id) {
            return *c.Name, string(c.LifecycleState)
		}
	}
	return "UNKNOWN", "UNKNOWN"
}
	
func display_formatted_list (parent_id string, level int, cpts []identity.Compartment) {
    // level = 0 for root, 1 for 1st level compartments, ...

	var cptname string
	var state string

    for i := 1; i < level; i++ {
        if flag[i] == 0 {
			fmt.Printf (COLOR_CYAN+"│      "+COLOR_NORMAL)
		} else {
            fmt.Printf ("       ")
		}
	}

    if level > 0 {
        cptname, state = get_cpt_name_and_state_from_id (parent_id, cpts)   
        if flag[level] == 0 {
			fmt.Printf (COLOR_CYAN+"├───── "+COLOR_NORMAL)
		} else {
            fmt.Printf (COLOR_CYAN+"└───── "+COLOR_NORMAL)
		}
    } else {
        cptname = "root"
        state   = "ACTIVE"
	}
	
    if state == "ACTIVE" {
		fmt.Println (COLOR_GREEN+cptname+COLOR_NORMAL+" "+parent_id+COLOR_YELLOW+" ACTIVE"+COLOR_NORMAL)
	} else {
        fmt.Println (COLOR_BLUE+cptname+COLOR_GREY+" "+parent_id+COLOR_RED+" DELETED"+COLOR_NORMAL)
	}

	// get the list of ids of the direct sub-compartments and store it in a Go slice
	slice := make([]string,0)
	for _, c := range cpts {
        if (*c.CompartmentId == parent_id) {
            slice = append (slice, *c.Id)
		}
	}
    
    // then for each of those cpt ids, display the sub-compartments details
	for i, cid := range slice {
		// if processing the last sub dir
		if i == len(slice)-1 {
			flag[level+1] = 1
		} else {
			flag[level+1] = 0
		}
		
		// display list of direct sub-folders
		display_formatted_list (cid, level+1, cpts)
	}
}

// -- main
func main() {
	
	// Check arguments passed
	if (len(os.Args) != 2) { usage() }
	profile := os.Args[1]

	// Try to load OCI config from profile
	config, err := common.ConfigurationProviderFromFileWithProfile(config_file, profile, "")
	client, err := identity.NewIdentityClientWithConfigurationProvider(config)
	helpers.FatalIfError(err)

	// Get tenancy OCID from profile
	tenancy_ocid, _ := config.TenancyOCID()

	// Get the list of all compartments and sub-comparments
	vrai := true
	request := identity.ListCompartmentsRequest{ 
		CompartmentId : common.String(tenancy_ocid), 
		CompartmentIdInSubtree : &vrai, 
	}
	list, err := client.ListCompartments(context.Background(), request)
	helpers.FatalIfError(err)

	// Display the list in a formatted output
	display_formatted_list (tenancy_ocid, 0, list.Items)
}
